File: python/voice_transcription_agent/voice_transcription_agent.py
"""
Voice Transcription Agent - Speech-to-Text conversion

This agent converts audio files or streams to text using various STT providers
Supports: OpenAI Whisper, Google Speech-to-Text, AssemblyAI, and more
"""

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from base.agent import BaseAgent
from typing import Optional, Dict, Any, Union
import tempfile


class VoiceTranscriptionAgent(BaseAgent):
    """
    A voice transcription agent that converts speech to text

    Features:
    - Multiple provider support (OpenAI Whisper, Google, AssemblyAI, etc.)
    - Audio file transcription
    - Real-time audio stream support
    - Language detection
    - Speaker diarization (provider-dependent)
    - Timestamp support
    - Custom vocabulary

    Example:
        agent = VoiceTranscriptionAgent({
            'provider': 'openai',  # or 'google', 'assemblyai'
            'api_key': 'your-api-key',
            'language': 'en',
            'model': 'whisper-1'
        })
        agent.initialize()

        # Transcribe a file
        result = agent.execute('audio.mp3')
        print(result['text'])

        # With timestamps
        result = agent.execute('audio.mp3', timestamps=True)
        for segment in result['segments']:
            print(f"{segment['start']}s - {segment['end']}s: {segment['text']}")
    """

    def _initialize(self) -> None:
        """Initialize voice transcription agent"""
        self.state['provider'] = self.config.get('provider', 'openai')
        self.state['api_key'] = self.config.get('api_key', os.getenv('TRANSCRIPTION_API_KEY'))
        self.state['language'] = self.config.get('language', 'en')
        self.state['model'] = self.config.get('model', 'whisper-1')

        # Provider-specific settings
        self.state['enable_diarization'] = self.config.get('enable_diarization', False)
        self.state['enable_timestamps'] = self.config.get('enable_timestamps', False)
        self.state['custom_vocabulary'] = self.config.get('custom_vocabulary', [])

        # Initialize provider client
        self._init_provider()

    def _init_provider(self) -> None:
        """Initialize the transcription provider client"""
        provider = self.state['provider']

        if provider == 'openai':
            # OpenAI Whisper setup
            try:
                import openai
                self.state['client'] = openai.OpenAI(api_key=self.state['api_key'])
                self.logger.info("OpenAI Whisper client initialized")
            except ImportError:
                self.logger.warning("OpenAI package not installed. Install with: pip install openai")

        elif provider == 'google':
            # Google Speech-to-Text setup
            try:
                from google.cloud import speech
                self.state['client'] = speech.SpeechClient()
                self.logger.info("Google Speech-to-Text client initialized")
            except ImportError:
                self.logger.warning("Google Cloud Speech package not installed. Install with: pip install google-cloud-speech")

        elif provider == 'assemblyai':
            # AssemblyAI setup
            try:
                import assemblyai as aai
                aai.settings.api_key = self.state['api_key']
                self.state['client'] = aai.Transcriber()
                self.logger.info("AssemblyAI client initialized")
            except ImportError:
                self.logger.warning("AssemblyAI package not installed. Install with: pip install assemblyai")

        elif provider == 'mock':
            self.state['client'] = None
            self.logger.info("Mock provider initialized")
        else:
            raise ValueError(f"Unsupported provider: {provider}")

    def execute(
        self,
        audio_input: Union[str, bytes],
        timestamps: bool = False,
        speaker_labels: bool = False,
        **kwargs
    ) -> Dict[str, Any]:
        """
        Transcribe audio to text

        Args:
            audio_input: Path to audio file or audio bytes
            timestamps: Include word/segment timestamps
            speaker_labels: Include speaker diarization
            **kwargs: Provider-specific parameters

        Returns:
            Dictionary containing:
            - text: Full transcription
            - language: Detected language (if available)
            - confidence: Confidence score
            - segments: Timestamped segments (if requested)
            - speakers: Speaker labels (if requested)
        """
        if not self._initialized:
            self.initialize()

        provider = self.state['provider']

        if provider == 'openai':
            return self._transcribe_openai(audio_input, timestamps, **kwargs)
        elif provider == 'google':
            return self._transcribe_google(audio_input, timestamps, speaker_labels, **kwargs)
        elif provider == 'assemblyai':
            return self._transcribe_assemblyai(audio_input, timestamps, speaker_labels, **kwargs)
        elif provider == 'mock':
            return self._transcribe_mock(audio_input)
        else:
            raise ValueError(f"Provider {provider} not supported")

    def _transcribe_openai(
        self,
        audio_input: Union[str, bytes],
        timestamps: bool = False,
        **kwargs
    ) -> Dict[str, Any]:
        """Transcribe using OpenAI Whisper API"""
        try:
            client = self.state['client']

            # Handle file path or bytes
            if isinstance(audio_input, str):
                audio_file = open(audio_input, 'rb')
            else:
                # Write bytes to temp file
                temp_file = tempfile.NamedTemporaryFile(delete=False, suffix='.mp3')
                temp_file.write(audio_input)
                temp_file.close()
                audio_file = open(temp_file.name, 'rb')

            # Transcribe
            response_format = 'verbose_json' if timestamps else 'text'

            transcript = client.audio.transcriptions.create(
                model=self.state['model'],
                file=audio_file,
                language=self.state['language'],
                response_format=response_format,
                **kwargs
            )

            audio_file.close()

            # Format response
            if timestamps and hasattr(transcript, 'segments'):
                return {
                    'text': transcript.text,
                    'language': getattr(transcript, 'language', self.state['language']),
                    'segments': [
                        {
                            'start': seg['start'],
                            'end': seg['end'],
                            'text': seg['text']
                        }
                        for seg in transcript.segments
                    ]
                }
            else:
                return {
                    'text': transcript if isinstance(transcript, str) else transcript.text,
                    'language': self.state['language']
                }

        except Exception as e:
            self.logger.error(f"OpenAI transcription error: {e}")
            raise

    def _transcribe_google(
        self,
        audio_input: Union[str, bytes],
        timestamps: bool = False,
        speaker_labels: bool = False,
        **kwargs
    ) -> Dict[str, Any]:
        """Transcribe using Google Speech-to-Text"""
        # Placeholder for Google implementation
        self.logger.info("Google Speech-to-Text transcription")

        # Implementation would be:
        # 1. Load audio file
        # 2. Configure recognition
        # 3. Call API
        # 4. Parse results

        raise NotImplementedError("Google Speech-to-Text integration pending. Add implementation here.")

    def _transcribe_assemblyai(
        self,
        audio_input: Union[str, bytes],
        timestamps: bool = False,
        speaker_labels: bool = False,
        **kwargs
    ) -> Dict[str, Any]:
        """Transcribe using AssemblyAI"""
        try:
            transcriber = self.state['client']

            config = {
                'speaker_labels': speaker_labels,
                'language_code': self.state['language']
            }

            if isinstance(audio_input, str):
                transcript = transcriber.transcribe(audio_input, config=config)
            else:
                # Save bytes to temp file
                temp_file = tempfile.NamedTemporaryFile(delete=False, suffix='.mp3')
                temp_file.write(audio_input)
                temp_file.close()
                transcript = transcriber.transcribe(temp_file.name, config=config)

            result = {
                'text': transcript.text,
                'confidence': transcript.confidence,
                'language': self.state['language']
            }

            if timestamps and hasattr(transcript, 'words'):
                result['segments'] = [
                    {
                        'start': word.start / 1000,  # Convert ms to seconds
                        'end': word.end / 1000,
                        'text': word.text,
                        'confidence': word.confidence
                    }
                    for word in transcript.words
                ]

            if speaker_labels and hasattr(transcript, 'utterances'):
                result['speakers'] = [
                    {
                        'speaker': utt.speaker,
                        'start': utt.start / 1000,
                        'end': utt.end / 1000,
                        'text': utt.text
                    }
                    for utt in transcript.utterances
                ]

            return result

        except Exception as e:
            self.logger.error(f"AssemblyAI transcription error: {e}")
            raise

    def _transcribe_mock(self, audio_input: Union[str, bytes]) -> Dict[str, Any]:
        """Mock transcription for testing"""
        filename = audio_input if isinstance(audio_input, str) else "audio_bytes"
        return {
            'text': f"Mock transcription of: {filename}",
            'language': self.state['language'],
            'confidence': 0.95
        }

    def transcribe_stream(self, audio_stream, callback=None):
        """
        Transcribe audio stream in real-time

        Args:
            audio_stream: Audio stream generator
            callback: Function to call with partial results

        Returns:
            Final transcription result
        """
        # This would implement real-time streaming transcription
        # Provider-dependent implementation
        raise NotImplementedError("Streaming transcription not yet implemented")

    def set_custom_vocabulary(self, words: list) -> None:
        """Set custom vocabulary for better recognition"""
        self.state['custom_vocabulary'] = words
        self.logger.info(f"Custom vocabulary set: {len(words)} words")


# Example usage
if __name__ == '__main__':
    print("Voice Transcription Agent Example")
    print("=" * 60)

    # Example 1: Mock provider (no API key needed)
    agent = VoiceTranscriptionAgent({
        'provider': 'mock',
        'language': 'en'
    })

    with agent:
        print("\n1. Basic transcription (mock):")
        result = agent.execute('example_audio.mp3')
        print(f"   Text: {result['text']}")
        print(f"   Language: {result['language']}")

    # Example 2: OpenAI Whisper (requires API key)
    print("\n2. OpenAI Whisper example (commented - requires API key):")
    print("""
    agent = VoiceTranscriptionAgent({
        'provider': 'openai',
        'api_key': 'your-api-key',
        'model': 'whisper-1'
    })

    with agent:
        # Simple transcription
        result = agent.execute('audio.mp3')
        print(result['text'])

        # With timestamps
        result = agent.execute('audio.mp3', timestamps=True)
        for segment in result['segments']:
            print(f"{segment['start']:.2f}s - {segment['end']:.2f}s: {segment['text']}")
    """)

    # Example 3: AssemblyAI with speaker diarization
    print("\n3. AssemblyAI example (commented - requires API key):")
    print("""
    agent = VoiceTranscriptionAgent({
        'provider': 'assemblyai',
        'api_key': 'your-api-key'
    })

    with agent:
        result = agent.execute(
            'meeting.mp3',
            timestamps=True,
            speaker_labels=True
        )

        # Print by speaker
        for speaker in result['speakers']:
            print(f"Speaker {speaker['speaker']}: {speaker['text']}")
    """)

File: javascript/voice-transcription-agent/VoiceTranscriptionAgent.js
/**
 * Voice Transcription Agent - Speech-to-Text conversion
 *
 * Supports: OpenAI Whisper, Google Speech-to-Text, AssemblyAI
 */

const BaseAgent = require('../base/Agent');
const fs = require('fs').promises;
const path = require('path');

class VoiceTranscriptionAgent extends BaseAgent {
  /**
   * Initialize voice transcription agent
   */
  async _initialize() {
    this.state.provider = this.config.provider || 'openai';
    this.state.apiKey = this.config.apiKey || process.env.TRANSCRIPTION_API_KEY;
    this.state.language = this.config.language || 'en';
    this.state.model = this.config.model || 'whisper-1';

    // Provider-specific settings
    this.state.enableDiarization = this.config.enableDiarization || false;
    this.state.enableTimestamps = this.config.enableTimestamps || false;
    this.state.customVocabulary = this.config.customVocabulary || [];

    // Initialize provider client
    await this._initProvider();
  }

  /**
   * Initialize the transcription provider client
   * @private
   */
  async _initProvider() {
    const provider = this.state.provider;

    if (provider === 'openai') {
      // OpenAI Whisper setup
      try {
        const OpenAI = require('openai');
        this.state.client = new OpenAI({ apiKey: this.state.apiKey });
        this.log('info', 'OpenAI Whisper client initialized');
      } catch (error) {
        this.log('warn', 'OpenAI package not installed. Install with: npm install openai');
      }
    } else if (provider === 'google') {
      // Google Speech-to-Text setup
      try {
        const speech = require('@google-cloud/speech');
        this.state.client = new speech.SpeechClient();
        this.log('info', 'Google Speech-to-Text client initialized');
      } catch (error) {
        this.log('warn', 'Google Cloud Speech package not installed. Install with: npm install @google-cloud/speech');
      }
    } else if (provider === 'assemblyai') {
      // AssemblyAI setup
      try {
        const assemblyai = require('assemblyai');
        assemblyai.settings.apiKey = this.state.apiKey;
        this.state.client = new assemblyai.Transcriber();
        this.log('info', 'AssemblyAI client initialized');
      } catch (error) {
        this.log('warn', 'AssemblyAI package not installed. Install with: npm install assemblyai');
      }
    } else if (provider === 'mock') {
      this.state.client = null;
      this.log('info', 'Mock provider initialized');
    } else {
      throw new Error(`Unsupported provider: ${provider}`);
    }
  }

  /**
   * Transcribe audio to text
   * @param {string|Buffer} audioInput - Path to audio file or audio buffer
   * @param {Object} options - Transcription options
   * @param {boolean} options.timestamps - Include timestamps
   * @param {boolean} options.speakerLabels - Include speaker diarization
   * @returns {Promise<Object>} Transcription result
   */
  async execute(audioInput, options = {}) {
    if (!this._initialized) {
      await this.initialize();
    }

    const { timestamps = false, speakerLabels = false, ...rest } = options;
    const provider = this.state.provider;

    if (provider === 'openai') {
      return await this._transcribeOpenAI(audioInput, timestamps, rest);
    } else if (provider === 'google') {
      return await this._transcribeGoogle(audioInput, timestamps, speakerLabels, rest);
    } else if (provider === 'assemblyai') {
      return await this._transcribeAssemblyAI(audioInput, timestamps, speakerLabels, rest);
    } else if (provider === 'mock') {
      return this._transcribeMock(audioInput);
    } else {
      throw new Error(`Provider ${provider} not supported`);
    }
  }

  /**
   * Transcribe using OpenAI Whisper API
   * @private
   */
  async _transcribeOpenAI(audioInput, timestamps, options) {
    try {
      const client = this.state.client;

      // Handle file path or buffer
      let audioBuffer;
      let filename;

      if (typeof audioInput === 'string') {
        audioBuffer = await fs.readFile(audioInput);
        filename = path.basename(audioInput);
      } else {
        audioBuffer = audioInput;
        filename = 'audio.mp3';
      }

      // Create File object for OpenAI API
      const audioFile = new File([audioBuffer], filename);

      // Transcribe
      const responseFormat = timestamps ? 'verbose_json' : 'text';

      const transcript = await client.audio.transcriptions.create({
        model: this.state.model,
        file: audioFile,
        language: this.state.language,
        response_format: responseFormat,
        ...options
      });

      // Format response
      if (timestamps && transcript.segments) {
        return {
          text: transcript.text,
          language: transcript.language || this.state.language,
          segments: transcript.segments.map(seg => ({
            start: seg.start,
            end: seg.end,
            text: seg.text
          }))
        };
      } else {
        return {
          text: typeof transcript === 'string' ? transcript : transcript.text,
          language: this.state.language
        };
      }
    } catch (error) {
      this.log('error', `OpenAI transcription error: ${error.message}`);
      throw error;
    }
  }

  /**
   * Transcribe using Google Speech-to-Text
   * @private
   */
  async _transcribeGoogle(audioInput, timestamps, speakerLabels, options) {
    // Placeholder for Google implementation
    this.log('info', 'Google Speech-to-Text transcription');

    throw new Error('Google Speech-to-Text integration pending. Add implementation here.');
  }

  /**
   * Transcribe using AssemblyAI
   * @private
   */
  async _transcribeAssemblyAI(audioInput, timestamps, speakerLabels, options) {
    try {
      const transcriber = this.state.client;

      const config = {
        speaker_labels: speakerLabels,
        language_code: this.state.language
      };

      const audioPath = typeof audioInput === 'string' ? audioInput : null;

      if (!audioPath) {
        throw new Error('AssemblyAI requires a file path');
      }

      const transcript = await transcriber.transcribe(audioPath, config);

      const result = {
        text: transcript.text,
        confidence: transcript.confidence,
        language: this.state.language
      };

      if (timestamps && transcript.words) {
        result.segments = transcript.words.map(word => ({
          start: word.start / 1000, // Convert ms to seconds
          end: word.end / 1000,
          text: word.text,
          confidence: word.confidence
        }));
      }

      if (speakerLabels && transcript.utterances) {
        result.speakers = transcript.utterances.map(utt => ({
          speaker: utt.speaker,
          start: utt.start / 1000,
          end: utt.end / 1000,
          text: utt.text
        }));
      }

      return result;
    } catch (error) {
      this.log('error', `AssemblyAI transcription error: ${error.message}`);
      throw error;
    }
  }

  /**
   * Mock transcription for testing
   * @private
   */
  _transcribeMock(audioInput) {
    const filename = typeof audioInput === 'string' ? audioInput : 'audio_buffer';
    return {
      text: `Mock transcription of: ${filename}`,
      language: this.state.language,
      confidence: 0.95
    };
  }

  /**
   * Transcribe audio stream in real-time
   * @param {ReadableStream} audioStream - Audio stream
   * @param {Function} callback - Callback for partial results
   * @returns {Promise<Object>} Final transcription result
   */
  async transcribeStream(audioStream, callback = null) {
    // This would implement real-time streaming transcription
    // Provider-dependent implementation
    throw new Error('Streaming transcription not yet implemented');
  }

  /**
   * Set custom vocabulary for better recognition
   * @param {Array<string>} words - Custom words/phrases
   */
  setCustomVocabulary(words) {
    this.state.customVocabulary = words;
    this.log('info', `Custom vocabulary set: ${words.length} words`);
  }
}

module.exports = VoiceTranscriptionAgent;

// Example usage
if (require.main === module) {
  (async () => {
    console.log('Voice Transcription Agent Example');
    console.log('='.repeat(60));

    // Example 1: Mock provider (no API key needed)
    const agent = new VoiceTranscriptionAgent({
      provider: 'mock',
      language: 'en'
    });

    await agent.use(async (agent) => {
      console.log('\n1. Basic transcription (mock):');
      const result = await agent.execute('example_audio.mp3');
      console.log(`   Text: ${result.text}`);
      console.log(`   Language: ${result.language}`);
    });

    // Example 2: OpenAI Whisper (requires API key)
    console.log('\n2. OpenAI Whisper example (commented - requires API key):');
    console.log(`
    const agent = new VoiceTranscriptionAgent({
      provider: 'openai',
      apiKey: 'your-api-key',
      model: 'whisper-1'
    });

    await agent.use(async (agent) => {
      // Simple transcription
      const result = await agent.execute('audio.mp3');
      console.log(result.text);

      // With timestamps
      const resultWithTime = await agent.execute('audio.mp3', { timestamps: true });
      for (const segment of resultWithTime.segments) {
        console.log(\`\${segment.start.toFixed(2)}s - \${segment.end.toFixed(2)}s: \${segment.text}\`);
      }
    });
    `);

    // Example 3: AssemblyAI with speaker diarization
    console.log('\n3. AssemblyAI example (commented - requires API key):');
    console.log(`
    const agent = new VoiceTranscriptionAgent({
      provider: 'assemblyai',
      apiKey: 'your-api-key'
    });

    await agent.use(async (agent) => {
      const result = await agent.execute(
        'meeting.mp3',
        { timestamps: true, speakerLabels: true }
      );

      // Print by speaker
      for (const speaker of result.speakers) {
        console.log(\`Speaker \${speaker.speaker}: \${speaker.text}\`);
      }
    });
    `);
  })();
}
